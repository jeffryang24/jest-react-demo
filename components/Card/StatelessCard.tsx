import { Card, Image } from 'semantic-ui-react';

interface Props {
  age: string;
  description: string;
  fullName: string;
}

export default function StatelessCard(props: Props) {
  return (
    <Card>
      <Image src="/matthew.png" wrapped ui={false} />
      <Card.Content>
        <Card.Header data-testid={'fullName'}>
          {props.fullName || 'No Name'}
        </Card.Header>
        <Card.Meta>
          <span>{'Age: '}</span>
          <span data-testid={'age'}>{props.age || 'No Age'}</span>
        </Card.Meta>
        <Card.Description data-testid={'description'}>
          {props.description || '-'}
        </Card.Description>
      </Card.Content>
    </Card>
  );
}
