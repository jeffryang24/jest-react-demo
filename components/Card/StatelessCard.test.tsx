import { ComponentPropsWithoutRef } from 'react';
import { render } from '@testing-library/react';

import StatelessCard from './StatelessCard';

type StatelessCardProps = ComponentPropsWithoutRef<typeof StatelessCard>;

describe('Components - StatelessCard', () => {
  function createProps(
    customProps?: Partial<StatelessCardProps>
  ): StatelessCardProps {
    return {
      age: '20',
      description: 'Lorem ipsum dolor sit amet',
      fullName: 'John Doe',
      ...customProps,
    };
  }

  function renderCard(props?: Partial<StatelessCardProps>) {
    const componentProps = createProps(props);
    return render(<StatelessCard {...componentProps} />);
  }

  // HINT: Use snapshot testing when you want to assert a complex result
  // which always have a same structure for every execution.
  it('should render the same result for the same props - snapshot', () => {
    const { baseElement } = renderCard();

    expect(baseElement).toMatchSnapshot();
  });

  it('should render fallback value when the props value are empty', () => {
    const { queryByTestId } = renderCard({
      age: '',
      description: '',
      fullName: '',
    });

    expect(queryByTestId('fullName')?.textContent).toBe('No Name');
    expect(queryByTestId('age')?.textContent).toBe('No Age');
    expect(queryByTestId('description')?.textContent).toBe('-');
  });

  it.todo('should skip this test since this test is still todo');
  it.skip('should skip this test since this test brokes the CI', () => {
    // Some flaky test here.
  });
});
