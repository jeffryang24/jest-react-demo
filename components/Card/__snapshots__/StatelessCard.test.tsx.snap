// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - StatelessCard should render the same result for the same props - snapshot 1`] = `
<body>
  <div>
    <div
      class="ui card"
    >
      <div
        class="image"
      >
        <img
          src="/matthew.png"
        />
      </div>
      <div
        class="content"
      >
        <div
          class="header"
          data-testid="fullName"
        >
          John Doe
        </div>
        <div
          class="meta"
        >
          <span>
            Age: 
          </span>
          <span
            data-testid="age"
          >
            20
          </span>
        </div>
        <div
          class="description"
          data-testid="description"
        >
          Lorem ipsum dolor sit amet
        </div>
      </div>
    </div>
  </div>
</body>
`;
